File: apps/desktop/src-tauri/native/system_audio.m
#import "system_audio.h"

#import <AVFoundation/AVFoundation.h>
#import <CoreMedia/CoreMedia.h>
#import <ScreenCaptureKit/ScreenCaptureKit.h>
#import <dispatch/dispatch.h>
#import <stdio.h>
#import <stdlib.h>
#import <string.h>

static nw_system_audio_callback_t g_callback = NULL;
static SCStream *g_stream = NULL;
static id<SCStreamOutput> g_output = NULL;
static dispatch_queue_t g_queue = NULL;

static BOOL nw_extract_float_samples(CMSampleBufferRef sampleBuffer, float **out, uint32_t *out_count) {
    if (CMSampleBufferGetNumSamples(sampleBuffer) == 0) {
        return NO;
    }

    CMBlockBufferRef blockBuffer = NULL;
    AudioBufferList audioBufferList;
    memset(&audioBufferList, 0, sizeof(audioBufferList));

    OSStatus status = CMSampleBufferGetAudioBufferListWithRetainedBlockBuffer(
        sampleBuffer,
        NULL,
        &audioBufferList,
        sizeof(audioBufferList),
        NULL,
        NULL,
        kCMSampleBufferFlag_AudioBufferList_Assure16ByteAlignment,
        &blockBuffer);

    if (status != noErr) {
        if (blockBuffer) {
            CFRelease(blockBuffer);
        }
        return NO;
    }

    AudioBufferList *mutableList = &audioBufferList;
    UInt32 bufferCount = mutableList->mNumberBuffers;

    if (bufferCount == 1) {
        AudioBuffer *buffer = &mutableList->mBuffers[0];
        if (buffer->mData == NULL || buffer->mDataByteSize == 0) {
            if (blockBuffer) {
                CFRelease(blockBuffer);
            }
            return NO;
        }
        uint32_t count = (uint32_t)(buffer->mDataByteSize / sizeof(float));
        float *copy = (float *)malloc(count * sizeof(float));
        if (copy == NULL) {
            if (blockBuffer) {
                CFRelease(blockBuffer);
            }
            return NO;
        }
        memcpy(copy, buffer->mData, count * sizeof(float));
        *out = copy;
        *out_count = count;
        if (blockBuffer) {
            CFRelease(blockBuffer);
        }
        return YES;
    }

    uint32_t maxSamples = 0;
    float **channels = (float **)calloc(bufferCount, sizeof(float *));
    uint32_t *channelCounts = (uint32_t *)calloc(bufferCount, sizeof(uint32_t));
    if (channels == NULL || channelCounts == NULL) {
        free(channels);
        free(channelCounts);
        if (blockBuffer) {
            CFRelease(blockBuffer);
        }
        return NO;
    }

    for (UInt32 i = 0; i < bufferCount; i++) {
        AudioBuffer *buffer = &mutableList->mBuffers[i];
        if (buffer->mData == NULL || buffer->mDataByteSize == 0) {
            continue;
        }
        uint32_t count = (uint32_t)(buffer->mDataByteSize / sizeof(float));
        float *copy = (float *)malloc(count * sizeof(float));
        if (copy == NULL) {
            for (UInt32 j = 0; j < i; j++) {
                free(channels[j]);
            }
            free(channels);
            free(channelCounts);
            if (blockBuffer) {
                CFRelease(blockBuffer);
            }
            return NO;
        }
        memcpy(copy, buffer->mData, count * sizeof(float));
        channels[i] = copy;
        channelCounts[i] = count;
        if (count > maxSamples) {
            maxSamples = count;
        }
    }

    if (maxSamples == 0) {
        free(channels);
        free(channelCounts);
        if (blockBuffer) {
            CFRelease(blockBuffer);
        }
        return NO;
    }

    uint32_t total = maxSamples * bufferCount;
    float *interleaved = (float *)calloc(total, sizeof(float));
    if (interleaved == NULL) {
        for (UInt32 i = 0; i < bufferCount; i++) {
            free(channels[i]);
        }
        free(channels);
        free(channelCounts);
        if (blockBuffer) {
            CFRelease(blockBuffer);
        }
        return NO;
    }

    for (uint32_t frame = 0; frame < maxSamples; frame++) {
        for (UInt32 ch = 0; ch < bufferCount; ch++) {
            float sample = 0.0f;
            if (channels[ch] != NULL && frame < channelCounts[ch]) {
                sample = channels[ch][frame];
            }
            interleaved[frame * bufferCount + ch] = sample;
        }
    }

    for (UInt32 i = 0; i < bufferCount; i++) {
        free(channels[i]);
    }
    free(channels);
    free(channelCounts);
    if (blockBuffer) {
        CFRelease(blockBuffer);
    }

    *out = interleaved;
    *out_count = total;
    return YES;
}

@interface NWSystemAudioOutput : NSObject <SCStreamOutput>
@end

@implementation NWSystemAudioOutput

- (void)stream:(SCStream *)stream
    didOutputSampleBuffer:(CMSampleBufferRef)sampleBuffer
                   ofType:(SCStreamOutputType)type {
    (void)stream;
    if (type != SCStreamOutputTypeAudio || g_callback == NULL) {
        return;
    }

    float *interleaved = NULL;
    uint32_t count = 0;
    if (!nw_extract_float_samples(sampleBuffer, &interleaved, &count) || count == 0) {
        return;
    }
    g_callback(interleaved, count, 48000, 2);
    free(interleaved);
}

@end

static void nw_start_capture(nw_system_audio_callback_t callback, void (^done)(int32_t code)) {
    if (g_stream != NULL) {
        done(0);
        return;
    }

    g_callback = callback;

    [SCShareableContent getShareableContentExcludingDesktopWindows:NO
                                             onScreenWindowsOnly:YES
                                               completionHandler:^(SCShareableContent *content, NSError *error) {
        if (error != nil) {
            fprintf(stderr, "SystemAudio: shareable content error: %s\n",
                    error.localizedDescription.UTF8String);
            done(-1);
            return;
        }

        SCDisplay *display = content.displays.firstObject;
        if (display == nil) {
            fputs("SystemAudio: no displays available\n", stderr);
            done(-2);
            return;
        }

        SCContentFilter *filter = [[SCContentFilter alloc] initWithDisplay:display excludingWindows:@[]];
        SCStreamConfiguration *config = [[SCStreamConfiguration alloc] init];
        config.capturesAudio = YES;
        config.excludesCurrentProcessAudio = YES;
        config.sampleRate = 48000;
        config.channelCount = 2;
        config.width = 2;
        config.height = 2;
        config.minimumFrameInterval = CMTimeMake(1, 1);

        NWSystemAudioOutput *output = [[NWSystemAudioOutput alloc] init];
        SCStream *stream = [[SCStream alloc] initWithFilter:filter configuration:config delegate:nil];

        if (g_queue == NULL) {
            g_queue = dispatch_queue_create("com.notewise.system-audio", DISPATCH_QUEUE_SERIAL);
        }

        NSError *addError = nil;
        if (![stream addStreamOutput:output type:SCStreamOutputTypeAudio sampleHandlerQueue:g_queue error:&addError]) {
            fprintf(stderr, "SystemAudio: addStreamOutput error: %s\n",
                    addError.localizedDescription.UTF8String);
            done(-4);
            return;
        }

        [stream startCaptureWithCompletionHandler:^(NSError *startError) {
            if (startError != nil) {
                fprintf(stderr, "SystemAudio: startCapture error: %s\n",
                        startError.localizedDescription.UTF8String);
                done(-3);
                return;
            }
            g_stream = stream;
            g_output = output;
            done(0);
        }];
    }];
}

static void nw_stop_capture(void (^done)(int32_t code)) {
    if (g_stream == NULL) {
        done(0);
        return;
    }

    SCStream *stream = g_stream;
    g_stream = NULL;
    g_output = NULL;
    g_callback = NULL;

    [stream stopCaptureWithCompletionHandler:^(NSError *error) {
        (void)error;
        done(0);
    }];
}

int32_t nw_system_audio_start(nw_system_audio_callback_t callback) {
    if (g_queue == NULL) {
        g_queue = dispatch_queue_create("com.notewise.system-audio", DISPATCH_QUEUE_SERIAL);
    }

    __block int32_t result = -99;
    dispatch_semaphore_t sem = dispatch_semaphore_create(0);

    dispatch_async(g_queue, ^{
        nw_start_capture(callback, ^(int32_t code) {
            result = code;
            dispatch_semaphore_signal(sem);
        });
    });

    dispatch_time_t timeout = dispatch_time(DISPATCH_TIME_NOW, (int64_t)(15 * NSEC_PER_SEC));
    dispatch_semaphore_wait(sem, timeout);
    return result;
}

int32_t nw_system_audio_stop(void) {
    if (g_queue == NULL) {
        return 0;
    }

    __block int32_t result = -99;
    dispatch_semaphore_t sem = dispatch_semaphore_create(0);

    dispatch_async(g_queue, ^{
        nw_stop_capture(^(int32_t code) {
            result = code;
            dispatch_semaphore_signal(sem);
        });
    });

    dispatch_time_t timeout = dispatch_time(DISPATCH_TIME_NOW, (int64_t)(5 * NSEC_PER_SEC));
    dispatch_semaphore_wait(sem, timeout);
    return result;
}

File: apps/desktop/src-tauri/native/system_audio.swift
import AVFoundation
import CoreMedia
import Foundation
import ScreenCaptureKit

private final class SystemAudioOutput: NSObject, SCStreamOutput {
    private let callback: (@convention(c) (UnsafePointer<Float>?, UInt32, UInt32, UInt16) -> Void)

    init(callback: @escaping @convention(c) (UnsafePointer<Float>?, UInt32, UInt32, UInt16) -> Void) {
        self.callback = callback
    }

    func stream(
        _ stream: SCStream,
        didOutputSampleBuffer sampleBuffer: CMSampleBuffer,
        of outputType: SCStreamOutputType
    ) {
        guard outputType == .audio else { return }
        guard let samples = Self.extractFloatSamples(from: sampleBuffer), !samples.isEmpty else { return }
        samples.withUnsafeBufferPointer { buffer in
            callback(buffer.baseAddress, UInt32(buffer.count), 48_000, 2)
        }
    }

    private static func extractFloatSamples(from sampleBuffer: CMSampleBuffer) -> [Float]? {
        guard CMSampleBufferGetNumSamples(sampleBuffer) > 0 else { return nil }

        var blockBuffer: CMBlockBuffer?
        var audioBufferList = AudioBufferList()
        let status = CMSampleBufferGetAudioBufferListWithRetainedBlockBuffer(
            sampleBuffer,
            bufferListSizeNeededOut: nil,
            bufferListOut: &audioBufferList,
            bufferListSize: MemoryLayout<AudioBufferList>.size,
            blockBufferAllocator: nil,
            blockBufferMemoryAllocator: nil,
            flags: kCMSampleBufferFlag_AudioBufferList_Assure16ByteAlignment,
            blockBufferOut: &blockBuffer
        )
        guard status == noErr else { return nil }

        let buffers = UnsafeMutableAudioBufferListPointer(&audioBufferList)
        if buffers.count == 1 {
            let buffer = buffers[0]
            guard let data = buffer.mData else { return nil }
            let count = Int(buffer.mDataByteSize) / MemoryLayout<Float>.size
            if count == 0 { return nil }
            let ptr = data.bindMemory(to: Float.self, capacity: count)
            return Array(UnsafeBufferPointer(start: ptr, count: count))
        }

        var channelData: [[Float]] = []
        var maxSamples = 0
        for buffer in buffers {
            guard let data = buffer.mData else { continue }
            let count = Int(buffer.mDataByteSize) / MemoryLayout<Float>.size
            if count == 0 { continue }
            let ptr = data.bindMemory(to: Float.self, capacity: count)
            channelData.append(Array(UnsafeBufferPointer(start: ptr, count: count)))
            maxSamples = max(maxSamples, count)
        }
        if channelData.isEmpty { return nil }

        var interleaved = [Float]()
        interleaved.reserveCapacity(maxSamples * channelData.count)
        for i in 0..<maxSamples {
            for channel in channelData {
                interleaved.append(i < channel.count ? channel[i] : 0)
            }
        }
        return interleaved
    }
}

private final class SystemAudioSession {
    static let shared = SystemAudioSession()

    private var stream: SCStream?
    private var output: SystemAudioOutput?
    private let queue = DispatchQueue(label: "com.notewise.system-audio", qos: .userInitiated)

    func start(
        callback: @escaping @convention(c) (UnsafePointer<Float>?, UInt32, UInt32, UInt16) -> Void,
        completion: @escaping (Int32) -> Void
    ) {
        queue.async {
            if self.stream != nil {
                completion(0)
                return
            }

            SCShareableContent.getExcludingDesktopWindows(false, onScreenWindowsOnly: true) { content, error in
                if let error {
                    fputs("SystemAudio: shareable content error: \(error.localizedDescription)\n", stderr)
                    completion(-1)
                    return
                }
                guard let display = content?.displays.first else {
                    fputs("SystemAudio: no displays available\n", stderr)
                    completion(-2)
                    return
                }

                let filter = SCContentFilter(display: display, excludingWindows: [])
                let config = SCStreamConfiguration()
                config.capturesAudio = true
                config.excludesCurrentProcessAudio = true
                config.sampleRate = 48_000
                config.channelCount = 2
                // ScreenCaptureKit requires a video surface — use a tiny dummy stream.
                config.width = 2
                config.height = 2
                config.minimumFrameInterval = CMTime(value: 1, timescale: 1)

                let output = SystemAudioOutput(callback: callback)
                let stream = SCStream(filter: filter, configuration: config, delegate: nil)
                do {
                    try stream.addStreamOutput(output, type: .audio, sampleHandlerQueue: self.queue)
                    stream.startCapture { startError in
                        if let startError {
                            fputs("SystemAudio: startCapture error: \(startError.localizedDescription)\n", stderr)
                            completion(-3)
                            return
                        }
                        self.stream = stream
                        self.output = output
                        completion(0)
                    }
                } catch {
                    fputs("SystemAudio: addStreamOutput error: \(error.localizedDescription)\n", stderr)
                    completion(-4)
                }
            }
        }
    }

    func stop(completion: @escaping (Int32) -> Void) {
        queue.async {
            guard let stream = self.stream else {
                completion(0)
                return
            }
            stream.stopCapture { _ in
                self.stream = nil
                self.output = nil
                completion(0)
            }
        }
    }
}

@_cdecl("nw_system_audio_start")
public func nw_system_audio_start(
    callback: @escaping @convention(c) (UnsafePointer<Float>?, UInt32, UInt32, UInt16) -> Void
) -> Int32 {
    let sem = DispatchSemaphore(value: 0)
    var result: Int32 = -99
    SystemAudioSession.shared.start(callback: callback) { code in
        result = code
        sem.signal()
    }
    _ = sem.wait(timeout: .now() + 15)
    return result
}

@_cdecl("nw_system_audio_stop")
public func nw_system_audio_stop() -> Int32 {
    let sem = DispatchSemaphore(value: 0)
    var result: Int32 = -99
    SystemAudioSession.shared.stop { code in
        result = code
        sem.signal()
    }
    _ = sem.wait(timeout: .now() + 5)
    return result
}
